apm-agent-plugins/apm-grails-plugin/src/main/java/co/elastic/apm/agent/grails/GrailsTransactionNameInstrumentation.java

```diff
@@ -0,0 +1,137 @@
+import co.elastic.apm.agent.bci.VisibleForAdvice;
+import co.elastic.apm.agent.impl.transaction.Transaction;
+import grails.core.GrailsControllerClass;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.grails.web.mapping.mvc.GrailsControllerUrlMappingInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.Collections;
+
+import static co.elastic.apm.agent.bci.bytebuddy.CustomElementMatchers.classLoaderCanLoadClass;
+import static co.elastic.apm.agent.impl.transaction.AbstractSpan.PRIO_HIGH_LEVEL_FRAMEWORK;
+import static grails.core.GrailsControllerClass.INDEX_ACTION;
+import static net.bytebuddy.matcher.ElementMatchers.hasSuperType;
+import static net.bytebuddy.matcher.ElementMatchers.isInterface;
+import static net.bytebuddy.matcher.ElementMatchers.nameStartsWith;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.not;
+import static net.bytebuddy.matcher.ElementMatchers.returns;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+
+public class GrailsTransactionNameInstrumentation extends ElasticApmInstrumentation {
+
+    @VisibleForAdvice
+    @SuppressWarnings("WeakerAccess")
+    public static final Logger logger = LoggerFactory.getLogger(GrailsTransactionNameInstrumentation.class);
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return nameStartsWith("org.grails.web.mapping.mvc")
+            .and(hasSuperType(named("org.springframework.web.servlet.HandlerAdapter")))
+            .and(not(isInterface()));
+    }
+
+    @Override
+    public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+        return named("handle")
+            .and(returns(named("org.springframework.web.servlet.ModelAndView")))
+            .and(takesArgument(0, named("javax.servlet.http.HttpServletRequest")))
+            .and(takesArgument(1, named("javax.servlet.http.HttpServletResponse")))
+            .and(takesArgument(2, Object.class));
+    }
+
+    @Override
+    public ElementMatcher.Junction<ClassLoader> getClassLoaderMatcher() {
+        return classLoaderCanLoadClass("org.grails.web.mapping.mvc.GrailsControllerUrlMappingInfo");
+    }
+
+    @Override
+    public Class<?> getAdviceClass() {
+        return HandlerAdapterAdvice.class;
+    }
+
+    @Override
+    public Collection<String> getInstrumentationGroupNames() {
+        return Collections.singletonList("grails");
+    }
+
+    @VisibleForAdvice
+    public static class HandlerAdapterAdvice {
+
+        @Advice.OnMethodEnter(suppress = Throwable.class)
+        static void setTransactionName(@Advice.Argument(2) Object handler) {
+            if (tracer != null) {
```

Comment:
We try to keep advice code minimal, thus moving most of the logic to `setName` method would be better here.

Two minor code style changes could be done here:
- Accessing logger directly from the advice could be removed as it does not provide much value (`null` transaction or tracer will not help to debug, but changing the transaction name will).
- using shortcut `return` statements helps to avoid numerous nested if statements, which helps readability (while it's not a very general "good code" practice, we tend to use it in advice code to make it easier to read).
- generating an exception to get the stack trace will likely not be required in production, thus we could remove it also.

Reply:
> We try to keep advice code minimal, thus moving most of the logic to setName method would be better here.

Not sure if that's possible in this case. We can only refer to Grails and Spring types within the advice.

I agree that the layout of the code and the loggers are not ideal but it's adapted from `co.elastic.apm.agent.spring.webmvc.SpringTransactionNameInstrumentation.HandlerAdapterAdvice#setTransactionName`. So I suggest to either fix both here or just do a followup with some cleanup, for consistency.

Reply:
I have removed the loggers and decreased nesting by adding early returns